[WebGPU] MTLResource allocations should be attributed to the respective web process

mwyrzykowski opened the pull request:
#### d9844889d42d79dc040865b5928b1da71fbdddf5
<pre>
[WebGPU] MTLResource allocations should be attributed to the respective web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=254219">https://bugs.webkit.org/show_bug.cgi?id=254219</a>
&lt;radar://107001772&gt;

Reviewed by Tadeu Zagallo and Myles C. Maxfield.

Attribute memory used by IOSurfaces and MTLResource instances to the
corresponding web process, so we do not exceed jetsam limits on iOS.

Mostly this involves piping the web process ID which already is
accessible via RemoteGPU down to WebGPU.framework and then calling the
appropriate IOSurface and Metal SPIs.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.cpp:
(PAL::WebGPU::create):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::safeCreateBuffer const):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::setOwnerWithIdentity):
(WebGPU::Device::setOwnerWithIdentity const):
(WebGPU::Device::newBufferWithBytes const):
(WebGPU::Device::newTextureWithDescriptor const):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::create):
(WebGPU::Instance::Instance):
(WebGPU::m_webProcessID):
(WebGPU::Instance::createSurface):
(WebGPU::Instance::webProcessID const):
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::create):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::renderBuffersWereRecreated):
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::webProcessID):
(WebKit::RemoteGPU::workQueueInitialize):

* Source/WebGPU/WebGPU/MetalSPI.h:
Add SPI header.

Canonical link: <a href="https://commits.webkit.org/276332@main">https://commits.webkit.org/276332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d700592f9641586ccdd5fe60866f1296b10f623

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23401 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46982 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46637 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27379 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20797 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44909 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20441 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38181 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17542 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17924 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39296 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2394 "Built successfully") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40523 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39576 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48593 "Built successfully") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43407 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20673 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42135 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9867 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21001 "Build is in progress. Recent messages:") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20298 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -189,6 +198,24 @@ static void captureFrame(id<MTLDevice> captureObject)
 
+static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
+{
+    ASSERT([resource respondsToSelector:@selector(setOwnerWithIdentity:)]);
```
The delegate method is optional hence the `respondsToSelector` check. But from what I can tell all resources we create implement it

kkinnunen-apple replied:
you can avoid the double respondsToSelector by structuring the statement as WebKit-style early return + ASSERT_NOT_REACHED()

litherum commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
```diff
@@ -96,7 +96,7 @@ class RemoteGPUProxy final : public PAL::WebGPU::GPU, private IPC::Connection::C
     IPC::Connection& connection() const { return m_gpuProcessConnection->connection(); }
 
-    void requestAdapter(const PAL::WebGPU::RequestAdapterOptions&, CompletionHandler<void(RefPtr<PAL::WebGPU::Adapter>&&)>&&) final;
+    void requestAdapter(const PAL::WebGPU::RequestAdapterOptions&, CompletionHandler<void(RefPtr<PAL::WebGPU::Adapter>&&)>&&, unsigned = 0) final;
```
What is the meaning of this? It needs a name. (And it seems unused in the .cpp file)

mwyrzykowski replied:
I fixed this up

litherum commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
```diff
@@ -118,7 +118,7 @@ void RemoteGPUProxy::waitUntilInitialized()
 }
 
-void RemoteGPUProxy::requestAdapter(const PAL::WebGPU::RequestAdapterOptions& options, CompletionHandler<void(RefPtr<PAL::WebGPU::Adapter>&&)>&& callback)
+void RemoteGPUProxy::requestAdapter(const PAL::WebGPU::RequestAdapterOptions& options, CompletionHandler<void(RefPtr<PAL::WebGPU::Adapter>&&)>&& callback, unsigned)
```
This doesn't seem used?

This indicates a leaky abstraction.

mwyrzykowski replied:
I will try to factor this away to avoid this issue

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
```diff
@@ -194,10 +196,16 @@ void RemoteGPU::createCompositorIntegration(WebGPUIdentifier identifier)
 
+unsigned RemoteGPU::webProcessID() const
+{
+    return m_gpuConnectionToWebProcess->webProcessIdentity().taskIdToken();
```
I think we should use the identity throughout our code, and only call `taskIdToken()` at the latest possible moment.

mwyrzykowski replied:
`ProcessIdentity` is in `WebCore/platform` so unfortunately I think this is the latest possible moment.

We can also forward declare `ProcessIdentity` and pass it throughout in PAL but then where do we use it?

litherum replied:
:/ ok

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPU.h:
```diff
@@ -42,7 +42,7 @@ class GPU : public RefCounted<GPU> {
     virtual ~GPU() = default;
 
-    virtual void requestAdapter(const RequestAdapterOptions&, CompletionHandler<void(RefPtr<Adapter>&&)>&&) = 0;
+    virtual void requestAdapter(const RequestAdapterOptions&, CompletionHandler<void(RefPtr<Adapter>&&)>&&, unsigned = 0) = 0;
```
Isn't there a better way to do this? Every `GPU` in the GPU process will come from a different web process, can't this information be in the `GPU` object (and not passed to requestAdapter())?

mwyrzykowski replied:
We could store it in the GPU object instead. This would require from what I can tell storing the GPU object in the Adapter or the Device so we can get the web process id.

Do you think storing a GPU& in the adapter is a good idea? I can do that if so

litherum replied:
I think that's the design we have already - child objects hold strong references to their parent object

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUCompositorIntegration.h:
```diff
@@ -44,6 +44,7 @@ class CompositorIntegration : public RefCounted<CompositorIntegration> {
 
 #if PLATFORM(COCOA)
     virtual Vector<MachSendRight> recreateRenderBuffers(int width, int height) = 0;
+    virtual void setWebProcessID(unsigned) { }
```
Right, this is a perfect example of why the Compositor Integration object exists - it's our own object with our own API that we fully control.

litherum commented on Source/WebGPU/WebGPU/Adapter.h:
```diff
@@ -64,16 +64,17 @@ class Adapter : public WGPUAdapterImpl, public RefCounted<Adapter> {
 
     Instance& instance() const { return m_instance; }
-
+    unsigned webProcessID() const { return m_webProcessID; }
```
I don't think this belongs here :/

mwyrzykowski replied:
oh maybe it should go in `Instance`? The name of that had me thinking there was only one `Instance` instance per GPU process. But it seems there is one `Instance` instance in the GPU process for each web process?

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -47,6 +47,15 @@
+#if USE(INTERNAL_APPLE_SDK)
+#import <Metal/MTLResource_Private.h>
+#else
+@protocol MTLResourceSPI <MTLResource>
```
Usually we put these in their own MetalSPI.h file

mwyrzykowski replied:
Should we introduce a new MetalSPI.h for WebGPU.framework? I don't think one exists yet

mwyrzykowski replied:
I didn't add one since I don't think a single SPI usage warrants its own file

litherum replied:
Yes, we should make a MetalSPI.h. It's important we keep accounting of SPI clear.

mwyrzykowski replied:
SPI header added :)

litherum commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -101,7 +101,10 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
-    return [m_device newBufferWithLength:std::max(static_cast<NSUInteger>(1), length) options:resourceOptions];
+    id<MTLBuffer> buffer = [m_device newBufferWithLength:std::max(static_cast<NSUInteger>(1), length) options:resourceOptions];
+    ASSERT(buffer);
+    setOwnerWithIdentity(buffer);
```
I don't understand, the `setOwnerWithIdentity()` function is on `Device` but I don't see a `m_device.` here?

mwyrzykowski replied:
We are inside `Device::safeCreateBuffer`

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -285,6 +312,22 @@ static void captureFrame(id<MTLDevice> captureObject)
+id<MTLBuffer> Device::newBufferWithBytes(const void *pointer, NSUInteger length, MTLResourceOptions options) const
+{
+    id<MTLBuffer> buffer = [m_device newBufferWithBytes:pointer length:length options:options];
+    ASSERT(buffer);
```
Is it really right to ASSERT() success? Don't we need to handle the case where allocation failed?

mwyrzykowski replied:
Yes we should handle that case. I'll file a bug to track that. The debug assert is to catch the failure when it happens as opposed to letting it go unnoticed.

litherum reviewed (changes requested):
It generally seems good, but I think we should rejigger the handshake between the GPU and the Device, so that we don't have to modify `pal/graphics/WebGPU/WebGPU.h`.

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -189,6 +198,24 @@ static void captureFrame(id<MTLDevice> captureObject)
+void Device::setOwnerWithIdentity(id<MTLResource> resource) const
+{
+    auto webProcessID = m_adapter->webProcessID();
+    if (!webProcessID)
```
I think this deserves a comment saying "WebKit1" or "In-process WebGPU" or something

litherum reviewed (approved):
I'm switching to approve because I think it's generally good, but I still think we should rejigger the handshake.

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.h:
```diff
@@ -86,12 +86,14 @@ class CompositorIntegrationImpl final : public CompositorIntegration {
 
     RefPtr<PresentationContextImpl> m_presentationContext;
     Ref<ConvertToBackingContext> m_convertToBackingContext;
+    unsigned m_webProcessID { 0 };
```
Isn't there a `mach_foo_t` type we can use instead of unsigned?

mwyrzykowski replied:
There is `mach_port_t` which is typedef as `unsigned int` but I don't think that is available in PAL. I could wrap all the usage of it in `#ifdef` however but this defeats the purpose of having a PAL 

I'm going to see if I can refactor some of these changes to have minimal impact on PAL.

mwyrzykowski replied:
I switched to `mach_port_t` as most PAL changes were removed. I think it will be fine, if there is a build issue on some platform, I will add a typedef

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
```diff
@@ -92,7 +104,7 @@ void RemoteGPU::workQueueInitialize()
     auto backing = PAL::WebGPU::create([protectedThis = Ref { *this }](PAL::WebGPU::WorkItem&& workItem) {
         protectedThis->workQueue().dispatch(WTFMove(workItem));
-    });
+    }, webProcessID(m_gpuConnectionToWebProcess));
```
navigating to gpuConnection is not thread safe. You need to store the process identity

mwyrzykowski replied:
Will storing the MachSendRight from the ProcessIdentity be sufficient? I store it in WebGPU.framework which doesn't have dependencies on WebCore.

kkinnunen-apple commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.cpp:
```diff
@@ -42,7 +42,7 @@ WTF_WEAK_LINK_FORCE_IMPORT(wgpuCreateInstance);
 namespace PAL::WebGPU {
 
-RefPtr<GPU> create(ScheduleWorkFunction&& scheduleWorkFunction)
+RefPtr<GPU> create(ScheduleWorkFunction&& scheduleWorkFunction, mach_port_t webProcessID)
```
It's probably not "web process id" if you are implementing a general purpose webgpu implementation.
I've been calling these "resource owner"
Being pedantic, the type is not `mach_port_t` but a `task_id_token_t`.

The value cannot be held willy-nilly, it's a refcounted object with a lifetime.
The best we have currently for holding this is WebCore::ProcessIdentity, if you can depend on WebCore::. Otherwise it would be WTF::MachSendRight.

mwyrzykowski replied:
I can't use WebCore in WebGPU.framework so I will store the MachSendRight

kkinnunen-apple commented on Source/WebGPU/WebGPU/MetalSPI.h:
```diff
@@ -0,0 +1,33 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if USE(INTERNAL_APPLE_SDK)
```
Not sure if it's relevant, but mentioning it anyway: It's unfortunate if the pal thingy at Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h cannot be used

mwyrzykowski replied:
Unfortunately I call the SPI method from WebGPU.framework. To use the one from WebCore I would need a way to get the Metal object from WebGPU.framework in WebCore, which is possible, but I think would side step some of the abstraction in place.

On the plus side having a separate header shows which minimal number of SPIs we are using in WebGPU.framework.

kkinnunen-apple commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -101,7 +101,11 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
     // FIXME(PERFORMANCE): Consider returning nil instead of clamping to 1.
     // FIXME(PERFORMANCE): Suballocate multiple Buffers either from MTLHeaps or from larger MTLBuffers.
-    return [m_device newBufferWithLength:std::max(static_cast<NSUInteger>(1), length) options:resourceOptions];
+    id<MTLBuffer> buffer = [m_device newBufferWithLength:std::max(static_cast<NSUInteger>(1), length) options:resourceOptions];
```
Just some ideas now that I'm looking at this, nothing to be concerned about:
you can also use `std::max<NSUinteger>(1, length)` to get less verbose maximum. 

You could avoid the inconvenient `ASSERT(buffer)` by just handling the allocation failure, same as before:
```
if (!buffer)
  return nil;
....
```
You will anyway write that code at some point when the fixme is fixed?

kkinnunen-apple commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2030,6 +2030,7 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonP
     // FIXME(PERFORMANCE): Consider implementing hazard tracking ourself.
 
     id<MTLTexture> texture = [m_device newTextureWithDescriptor:textureDescriptor];
+    setOwnerWithIdentity(texture);
```
does not make sense with the check below

kkinnunen-apple commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -112,6 +112,8 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
 
     uint32_t maxBuffersPlusVertexBuffersForVertexStage() const;
     uint32_t vertexBufferIndexForBindGroup(uint32_t groupIndex) const;
+    id<MTLBuffer> newBufferWithBytes(const void *, NSUInteger, MTLResourceOptions) const;
```
Style-wise, I think the new hunks have wrong position for the star in pointers

Might also in the future be easier to ingest size_t, as most of callers seem to do casting to NSUInteger.

mattwoodrow commented on Source/WebGPU/WebGPU/Instance.h:
```diff
@@ -61,16 +65,18 @@ class Instance : public WGPUInstanceImpl, public ThreadSafeRefCounted<Instance>
 
     // This can be used on a background thread.
     Deque<WGPUWorkItem> m_pendingWork WTF_GUARDED_BY_LOCK(m_lock);
+    const WTF::MachSendRight* m_webProcessResourceOwner { nullptr };
```
We shouldn't have a raw pointer member here, can't we just copy it?

mwyrzykowski replied:
oh copying is a good idea, I will do that

mattwoodrow commented on Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
```diff
@@ -65,6 +66,9 @@ class PresentationContextIOSurface : public PresentationContext {
     id<MTLFunction> m_luminanceClampFunction;
     id<MTLComputePipelineState> m_computePipelineState;
+#if HAVE(IOSURFACE_SET_OWNERSHIP_IDENTITY) && HAVE(TASK_IDENTITY_TOKEN)
+    const Instance& m_instance;
```
Raw reference here too, should at least be CheckedRef or something. How do we ensure all of these are destroyed before the creating `Instance`?

mattwoodrow commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1042,9 +1042,7 @@ RefPtr<WebCore::WebGPU::GPU> WebChromeClient::createGPUForWebGPU() const
-    return WebCore::WebGPU::create([](WebCore::WebGPU::WorkItem&& workItem) {
-        callOnMainRunLoop(WTFMove(workItem));
-    });
+    return nullptr;
```
Does this remove support for running WebGPU in WK1/WebProcess?

If it previously somewhat worked (at least for testing), then it'd be good to have this mentioned in the commit message (even in the title, if someone might have been using it).

mwyrzykowski replied:
We haven’t supported WebGPU in WK1 for over a year and we don’t have plans to add support for it

litherum replied:
😭

mwyrzykowski replied:
oh looks like @achristensen07 wrote some tests which use WebGPU and WK1. I will put this back but generally it is not supported

mwyrzykowski committed:
[WebGPU] MTLResource allocations should be attributed to the respective web process
https://bugs.webkit.org/show_bug.cgi?id=254219
<radar://107001772>

Reviewed by Tadeu Zagallo and Myles C. Maxfield.

Attribute memory used by IOSurfaces and MTLResource instances to the
corresponding web process, so we do not exceed jetsam limits on iOS.

Mostly this involves piping the web process ID which already is
accessible via RemoteGPU down to WebGPU.framework and then calling the
appropriate IOSurface and Metal SPIs.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.cpp:
(PAL::WebGPU::create):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::safeCreateBuffer const):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::setOwnerWithIdentity):
(WebGPU::Device::setOwnerWithIdentity const):
(WebGPU::Device::newBufferWithBytes const):
(WebGPU::Device::newTextureWithDescriptor const):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::create):
(WebGPU::Instance::Instance):
(WebGPU::m_webProcessID):
(WebGPU::Instance::createSurface):
(WebGPU::Instance::webProcessID const):
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::create):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::renderBuffersWereRecreated):
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::webProcessID):
(WebKit::RemoteGPU::workQueueInitialize):

* Source/WebGPU/WebGPU/MetalSPI.h:
Add SPI header.

Canonical link: https://commits.webkit.org/276332@main
